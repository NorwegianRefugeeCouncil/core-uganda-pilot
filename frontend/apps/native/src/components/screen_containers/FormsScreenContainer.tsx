import { FormDefinition } from 'core-api-client';
import React from 'react';

import { FormsScreenContainerProps } from '../../types/screens';
import useApiClient from '../../utils/clients';
import { FormsScreen } from '../screens/FormsScreen';

export const FormsScreenContainer = ({
  navigation,
  route,
}: FormsScreenContainerProps) => {
  const [forms, setForms] = React.useState<FormDefinition[]>();
  const [isLoading, setIsLoading] = React.useState(true);
  const apiClient = useApiClient();

  React.useEffect(() => {
    apiClient.Form.list({}).then((data) => {
      setIsLoading(false);
      setForms(data.response?.items);
    });
  }, []);
  return (
    <FormsScreen isLoading={isLoading} forms={forms} navigation={navigation} />
  );
};
