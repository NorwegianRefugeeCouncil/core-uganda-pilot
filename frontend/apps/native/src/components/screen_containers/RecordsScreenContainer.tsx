import React from 'react';

import { RECORD_ACTIONS } from '../../reducers/recordsReducers';
import { RecordsScreenContainerProps } from '../../types/screens';
import RecordsScreen from '../screens/RecordsScreen';
import client from '../../utils/clients';

export const RecordsScreenContainer = ({ navigation, route, state, dispatch }: RecordsScreenContainerProps) => {
  const { formId, databaseId } = route.params;
  const [isLoading, setIsLoading] = React.useState(true);

  React.useEffect(() => {
    if (formId && databaseId) {
      client()
        .listRecords({ formId, databaseId })
        .then((data) => {
          dispatch({
            type: RECORD_ACTIONS.GET_RECORDS,
            payload: {
              formId,
              records: data.response?.items,
            },
          });
        })
        .catch(console.error)
        .finally(() => setIsLoading(false));
    }
  }, [client, formId, databaseId]);

  const recordsScreenProps = { isLoading, state, navigation, route };
  return <RecordsScreen {...recordsScreenProps} />;
};
