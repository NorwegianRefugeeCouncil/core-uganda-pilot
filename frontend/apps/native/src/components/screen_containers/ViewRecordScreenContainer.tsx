import React from "react";
import ViewRecordScreen, {ViewRecordScreenProps} from "../screens/ViewRecordScreen";
import {ViewRecordScreenContainerProps} from "../../types/screens";
import client from "../../utils/clients";
import {useForm} from "react-hook-form";
import {FormDefinition} from "core-api-client";

export const ViewRecordScreenContainer = ({route, state}: ViewRecordScreenContainerProps) => {
    const {formId, recordId} = route.params;

    const [isLoading, setIsLoading] = React.useState(true);
    const [form, setForm] = React.useState<FormDefinition>();

    const {control, reset} = useForm();

    React.useEffect(() => {
        const getForm = async () => {
            try {
                const {response} = await client().getForm({id: formId});
                setForm(response);
            } catch (err) {
                console.error(err);
            }
        };
        getForm();
    }, [formId]);

    React.useEffect(() => {
        if (form) {
            reset(state.formsById[formId].recordsById[recordId].values);
            setIsLoading(false);
        }
    }, [form]);

    const viewRecordScreenProps: ViewRecordScreenProps = {
        isLoading,
        form,
        control,
    };
    return <ViewRecordScreen {...viewRecordScreenProps} />;
};
