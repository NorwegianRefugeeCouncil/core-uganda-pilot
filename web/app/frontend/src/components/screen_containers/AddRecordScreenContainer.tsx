import { FormDefinition } from "core-js-api-client";
import { NetworkStateType } from "expo-network";
import React from "react";
import { useForm } from "react-hook-form";
import { Platform } from "react-native";
import { RECORD_ACTIONS } from "../../reducers/recordsReducers";
import { AddRecordScreenContainerProps } from "../../types/screens";
import client from "../../utils/clients";
import { getEncryptionKey } from "../../utils/getEncryptionKey";
import { getNetworkState } from "../../utils/getNetworkState";
import { getEncryptedLocalData, storeEncryptedLocalData } from "../../utils/storage";
import { AddRecordScreen } from "../screens/AddRecordScreen";

export const AddRecordScreenContainer = ({ route, dispatch }: AddRecordScreenContainerProps) => {
    const { formId, recordId } = route.params;

    const isWeb = Platform.OS === "web";

    const [isLoading, setIsLoading] = React.useState(true);
    const [form, setForm] = React.useState<FormDefinition>();
    const [isConnected, setIsConnected] = React.useState(false);
    const [hasLocalData, setHasLocalData] = React.useState(false);

    const { control, handleSubmit, formState, reset } = useForm();

    React.useEffect(() => {
        async function fetches() {
            let form, localData, networkState;

            // react to network changes
            try {
                networkState = await getNetworkState();
            } catch (error) {
                console.error(error);
                setIsLoading(true);
            } finally {
                setIsConnected(networkState !== NetworkStateType.NONE);
            }

            //
            try {
                const data = await client().getForm({ id: formId });
                form = data?.response;
            } catch (error) {
                console.error(error);
                setIsLoading(true);
            } finally {
                setForm(form);
                setIsLoading(false);
            }

            // check for locally stored data on mobile device
            if (!isWeb && recordId) {
                try {
                    localData = await getEncryptedLocalData(recordId);
                } catch (error) {
                    console.error(error);
                } finally {
                    setHasLocalData(!!localData);
                    reset(localData);
                }
            }
        }
        fetches();
    });

    const onSubmitOffline = async (data: any) => {
        const key = getEncryptionKey();
        try {
            await storeEncryptedLocalData(recordId, key, data);
            dispatch({
                type: RECORD_ACTIONS.ADD_LOCAL_RECORD,
                payload: {
                    formId,
                    localRecord: recordId,
                },
            });
            setHasLocalData(true);
        } catch (e) {
            setHasLocalData(false);
        }
    };

    const onSubmit = (data: any) => {
        handleSubmit(async () => {
            if (isConnected || isWeb) {
                await client().createRecord({
                    object: { formId, values: data },
                });
            } else {
                await onSubmitOffline(data);
            }
        });
    };

    return (
        <AddRecordScreen
            form={form}
            control={control}
            onSubmit={onSubmit}
            formState={formState}
            isWeb={isWeb}
            hasLocalData={hasLocalData}
            isConnected={isConnected}
            isLoading={isLoading}
        />
    );
};
