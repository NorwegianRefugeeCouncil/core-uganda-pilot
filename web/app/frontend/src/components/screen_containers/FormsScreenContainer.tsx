import { FormDefinition } from "core-js-api-client/lib/types/types";
import React from "react";

import { FormsScreenContainerProps } from "../../types/screens";
import client from "../../utils/clients";
import { FormsScreen } from "../screens/FormsScreen";

export const FormsScreenContainer = ({
    navigation,
    route,
}: FormsScreenContainerProps) => {
    const [forms, setForms] = React.useState<FormDefinition[]>();
    const [isLoading, setIsLoading] = React.useState(true);
    React.useEffect(() => {
        client.listForms({}).then((data) => {
            setIsLoading(false);
            setForms(data.response?.items);
        });
    });
    return (
        <FormsScreen
            isLoading={isLoading}
            forms={forms}
            navigation={navigation}
        />
    );
};
