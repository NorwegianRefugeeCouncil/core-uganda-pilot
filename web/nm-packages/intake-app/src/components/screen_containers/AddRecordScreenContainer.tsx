import { FormDefinition } from 'core-js-api-client/lib/types/types';
import { NetworkStateType } from 'expo-network';
import React from 'react';
import { useForm } from 'react-hook-form';
import { Platform } from 'react-native';

import { RECORD_ACTIONS } from '../../reducers/recordsReducers';
import { AddRecordScreenContainerProps } from '../../types/screens';
import { getEncryptionKey } from '../../utils/getEncryptionKey';
import { getNetworkState } from '../../utils/getNetworkState';
import {
    getEncryptedLocalData,
    storeEncryptedLocalData,
} from '../../utils/storage';
import { useApiClient } from '../../utils/useApiClient';
import { AddRecordScreen } from '../screens/AddRecordScreen';

export const AddRecordScreenContainer = ({
    route,
    dispatch,
}: AddRecordScreenContainerProps) => {
    const { formId, recordId } = route.params;

    const client = useApiClient();
    const isWeb = Platform.OS === 'web';

    const [isLoading, setIsLoading] = React.useState(true);
    const [form, setForm] = React.useState<FormDefinition>();
    // const [simulateOffline, setSimulateOffline] = React.useState(!isWeb); // TODO: for testing, remove
    const [isConnected, setIsConnected] = React.useState(false);
    // const [showSnackbar, setShowSnackbar] = React.useState(!isConnected);
    const [hasLocalData, setHasLocalData] = React.useState(false);

    const { control, handleSubmit, formState, reset } = useForm();

    React.useEffect(() => {
        async function fetches() {
            let form, localData, networkState;

            // react to network changes
            try {
                networkState = await getNetworkState();
            } catch (error) {
                console.error(error);
                setIsLoading(true);
            } finally {
                setIsConnected(networkState === NetworkStateType.NONE);
            }

            //
            try {
                const data = await client.getForm({ id: formId });
                form = data?.response;
            } catch (error) {
                console.error(error);
                setIsLoading(true);
            } finally {
                setForm(form);
                setIsLoading(false);
            }

            // check for locally stored data on mobile device
            if (!isWeb && recordId) {
                try {
                    localData = await getEncryptedLocalData(recordId);
                } catch (error) {
                    console.error(error);
                } finally {
                    setHasLocalData(!!localData);
                    reset(localData);
                }
            }
        }
        fetches();
    });

    const onSubmitOffline = async (data: any) => {
        const key = getEncryptionKey();

        storeEncryptedLocalData(recordId, key, data)
            .then(() => {
                setHasLocalData(true);
                dispatch({
                    type: RECORD_ACTIONS.ADD_LOCAL_RECORD,
                    payload: {
                        formId,
                        localRecord: recordId,
                    },
                });
            })
            .catch(() => {
                setHasLocalData(false);
            });
    };

    const onSubmit = (data: any) => {
        handleSubmit(async () => {
            if (isConnected || isWeb) {
                await client.createRecord({ object: { formId, values: data } });
            } else {
                await onSubmitOffline(data);
            }
        });
    };

    return (
        <AddRecordScreen
            form={form}
            control={control}
            onSubmit={onSubmit}
            formState={formState}
            isWeb={isWeb}
            hasLocalData={hasLocalData}
            isConnected={isConnected}
            isLoading={isLoading}
        />
    );
};
