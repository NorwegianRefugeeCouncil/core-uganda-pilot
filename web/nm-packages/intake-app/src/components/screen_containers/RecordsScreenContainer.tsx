import React from 'react';
import { RecordsScreenContainerProps } from '../../types/screens';
import {RecordsScreen} from '../screens/RecordsScreen';
import { useApiClient } from '../../utils/useApiClient';
import { RECORD_ACTIONS } from '../../reducers/recordsReducers';

export const RecordsScreenContainer = ({
    navigation,
    route,
    state,
    dispatch,
}: RecordsScreenContainerProps) => {
    const client = useApiClient();
    const { formId, databaseId } = route.params;
    const [isLoading, setIsLoading] = React.useState(true);
    const { records, localRecords } = state.formsById[formId];

    React.useEffect(() => {
        client.listRecords({ formId, databaseId }).then(data => {
            dispatch({
                type: RECORD_ACTIONS.GET_RECORDS,
                payload: {
                    formId,
                    records: data.response?.items,
                },
            });
            setIsLoading(false);
        });
    }, []);
    return (
        <RecordsScreen
            navigation={navigation}
            isLoading={isLoading}
            records={records}
            localRecords={localRecords}
            formId={formId}
        />
    );
};
