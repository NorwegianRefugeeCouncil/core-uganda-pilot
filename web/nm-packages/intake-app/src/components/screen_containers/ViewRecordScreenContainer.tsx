import React from 'react';
import {ViewRecordScreenContainerProps} from '../../types/screens';
import {FormDefinition} from 'core-js-api-client/lib/types/types';
import {useApiClient} from '../../utils/useApiClient';
import {useForm} from 'react-hook-form';
import {ViewRecordScreen} from '../screens/ViewRecordScreen';

export const ViewRecordScreenContainer = ({
                                              route,
                                              state,
                                          }: ViewRecordScreenContainerProps) => {
    const {formId, recordId} = route.params;

    const [isLoading, setIsLoading] = React.useState(true);
    const [form, setForm] = React.useState<FormDefinition>();

    const client = useApiClient();
    const {control, reset} = useForm();

    React.useEffect(() => {
        client.getForm({id: formId}).then(data => {
            setForm(data.response);
        });
    }, [formId]);

    React.useEffect(() => {
        if (form) {
            reset(state.formsById[formId].recordsById[recordId].values);
            setIsLoading(false);
        }
    }, [form]);

    return <ViewRecordScreen isLoading={isLoading} control={control}/>;
};
