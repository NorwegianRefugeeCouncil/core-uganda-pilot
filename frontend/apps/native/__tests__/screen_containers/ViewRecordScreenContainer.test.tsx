import { render, waitFor } from '@testing-library/react-native';
import React from 'react';
import { FormDefinition } from 'core-js-api-client';

import { ViewRecordScreenContainer } from '../../src/components/screen_containers/ViewRecordScreenContainer';

const mockFormId = 'a form id';
const mockRecordId = 'a record id';

const mockRoute = {
  params: {
    formId: mockFormId,
    recordId: mockRecordId,
  },
};

const mockState: any = {
  formsById: {},
};

mockState.formsById[mockFormId] = {
  recordsById: {},
};

const mockValue = {
  fieldId: 'a field id',
  value: 'a value',
};
const mockValues = Array(10).fill(mockValue);

mockState.formsById[mockFormId].recordsById[mockRecordId] = { values: mockValues };

const mockForm: FormDefinition = {
  id: mockFormId,
  code: 'a code',
  databaseId: 'a database id',
  folderId: 'a folder id',
  name: 'a name',
  fields: [],
};

const mockFormData: { response: FormDefinition } = {
  response: mockForm,
};
const mockGetForm = jest.fn().mockResolvedValue(mockFormData);

jest.mock('../../src/utils/clients', () => () => ({
  getForm: mockGetForm,
}));

const mockControl = {};
const mockReset = jest.fn();
jest.mock('react-hook-form', () => {
  const rhf = jest.requireActual('react-hook-form');
  return {
    ...rhf,
    useForm: () => ({
      control: mockControl,
      reset: mockReset,
    }),
  };
});

const mockProps: any = {
  route: mockRoute,
  state: mockState,
};

describe(ViewRecordScreenContainer.name, () => {
  test('renders correctly', async () => {
    const { toJSON } = await waitFor(() => render(<ViewRecordScreenContainer {...mockProps} />));
    expect(toJSON()).toMatchSnapshot();
  });

  test('fetches a form', async () => {
    await waitFor(() => render(<ViewRecordScreenContainer {...mockProps} />));

    const expectedGetFormArg = { id: mockFormId };
    expect(mockGetForm).toHaveBeenCalledWith(expectedGetFormArg);
  });

  test("call react-hook-form's reset method with form data", async () => {
    await waitFor(() => render(<ViewRecordScreenContainer {...mockProps} />));
    await waitFor(() => expect(mockReset).toHaveBeenCalledWith(mockValues));
  });
});
