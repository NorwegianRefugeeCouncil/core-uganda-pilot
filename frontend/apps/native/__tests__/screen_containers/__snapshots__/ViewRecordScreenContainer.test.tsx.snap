// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewRecordScreenContainer renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
        "flexGrow": 1,
      },
      Object {
        "alignItems": "flex-start",
        "flexGrow": 1,
        "justifyContent": "flex-start",
        "padding": 18,
      },
      Object {
        "backgroundColor": "#24303E",
        "color": "#000000",
      },
    ]
  }
>
  <RCTScrollView>
    <View>
      <View />
    </View>
  </RCTScrollView>
</View>
`;
